app/forms/course_search_form.rb

```diff
@@ -0,0 +1,20 @@
+# frozen_string_literal: true
+
+class CourseSearchForm
```

Comment:
I'm not sure I see the value in this form object.

It looks to filter down the object keys from query params, but that's all. Argument/parameter validation feels to me that it should be the responsibility of the query object itself?

Reply:
I am not sure either the transformation from raw parameters into ruby objects (booleans, arrays, strings, value conversions, etc) should be in the query class. 

I know you wanna avoid indirection but then the Query class would have too much responsibility.

🤔 🤔 🤔 

In my opinion, the best we could do if we remove the form object and make the code clean:

```
class CoursesQuery
  class Params
    include ActiveModel::Attributes

    attribute :can_sponsor_visa, :boolean
  end

  def self.call(params = {})
    new(params:).call
  end

  attr_reader :scope, :params

  def initialize(params:)
    @params = CoursesQuery::Params.new(params)
```

Then in the future we could do something like:

```ruby
class Params
    include ActiveModel::Model
    include ActiveModel::Attributes

    attribute :can_sponsor_visa, :boolean
    attribute :age_group, :string
    attribute :applications_open, :boolean
    attribute :subjects, :array, default: []

    # Normalization and legacy param handling
    def initialize(params = {})
      super(normalize_params(params))
    end

    private

    def normalize_params(params)
      params.dup.tap do |p|
        p[:can_sponsor_visa] = p.delete(:visa_status) if p[:visa_status].present?
        p[:subjects] = Array(p[:subjects]).transform # transform subjects somehow
      end
    end
  end
```

Wdyt?

Reply:
I'm not sure we need to do that either.

Let's assume params are directly consumed from the query params, from the following query:

```
?can_sponsor_visa=true&age_group=abc&applications_open=T&subjects[]=English&subjects[]=Mathematics
```

We can do:

```ruby
class CoursesQuery
  # Handling boolean params - Cast if not blank before usage
  # "applications_open" param would also work the same
  def sponsor_visa_filter
    return @scope if params[:can_sponsor_visa].blank?

    can_sponsor_visa = ActiveModel::Type::Boolean.cast(params[:can_sponsor_visa])

    @scope.where(can_sponsor_visa:)
  end

  # Allowlist/enum filter
  def age_group_filter
    return @scope unless params[:age_group].in?(%w[some list of valid values])

    @scope.where(age_group: params[:age_group])
  end
end
```

This way, the params are handled only if they are used and ignored if invalid/unused.

Reply:
I am not sure I agree but I think for the visa sponsorship only, it makes total sense for now. I will implement.